src/backend/distributed/utils/metadata_cache.c

```diff
@@ -677,6 +678,25 @@ InvalidateDistRelationCacheCallback(Datum argument, Oid relationId)
 			cacheEntry->isValid = false;
 		}
 	}
+
+	/*
+	 * If pg_dist_partition is being invalidated drop all state
+	 * This happens pretty rarely, but most importantly happens during
+	 * DROP EXTENSION citus;
+	 */
+
+	if (relationId != InvalidOid && relationId == distPartitionRelationId)
```

Comment:
I like the separation, this method has two sections and each has a different responsibility.

Also, if relationId is simply not null we want to invalidate the cache entry. If relationId is not null and also distPartitionRelationID we want to both invalidate the cache entry and drop the rest of the state. I think expressing that is cleaner with two if blocks.